Python/optimizer_symbols.c

```diff
@@ -48,187 +43,250 @@ static inline int get_lltrace(void) {
 {
-    return sym->const_val;
+    if (sym->tag == JIT_SYM_KNOWN_VALUE_TAG) {
+        return sym->value.value;
+    }
+    return NULL;
 }
 
 void
-_Py_uop_sym_set_type(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym, PyTypeObject *typ)
+_Py_uop_sym_set_type(JitOptContext *ctx, JitOptSymbol *sym, PyTypeObject *typ)
 {
-    assert(typ != NULL && PyType_Check(typ));
-    if (sym->flags & IS_NULL) {
-        sym_set_bottom(ctx, sym);
-        return;
-    }
-    if (sym->typ != NULL) {
-        if (sym->typ != typ) {
+    JitSymType tag = sym->tag;
+    switch(tag) {
+        case JIT_SYM_NULL_TAG:
             sym_set_bottom(ctx, sym);
             return;
-        }
-    }
-    else {
-        sym_set_flag(sym, NOT_NULL);
-        sym->typ = typ;
+        case JIT_SYM_KNOWN_CLASS_TAG:
+            if (sym->cls.type != typ) {
+                sym_set_bottom(ctx, sym);
+            }
+            return;
+        case JIT_SYM_TYPE_VERSION_TAG:
+            if (sym->version.version == typ->tp_version_tag) {
+                sym->tag = JIT_SYM_KNOWN_CLASS_TAG;
+                sym->cls.type = typ;
+                sym->cls.version = typ->tp_version_tag;
+            }
+            else {
+                sym_set_bottom(ctx, sym);
+            }
+            return;
+        case JIT_SYM_KNOWN_VALUE_TAG:
+            if (Py_TYPE(sym->value.value) != typ) {
+                Py_CLEAR(sym->value.value);
+                sym_set_bottom(ctx, sym);
+            }
+            return;
+        case JIT_SYM_TUPLE_TAG:
+            if (typ != &PyTuple_Type) {
+                sym_set_bottom(ctx, sym);
+            }
+            return;
+        case JIT_SYM_BOTTOM_TAG:
+            return;
+        case JIT_SYM_NON_NULL_TAG:
+        case JIT_SYM_UNKNOWN_TAG:
+            sym->tag = JIT_SYM_KNOWN_CLASS_TAG;
+            sym->cls.version = 0;
+            sym->cls.type = typ;
+            return;
     }
 }
 
 bool
-_Py_uop_sym_set_type_version(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym, unsigned int version)
+_Py_uop_sym_set_type_version(JitOptContext *ctx, JitOptSymbol *sym, unsigned int version)
 {
-    // if the type version was already set, then it must be different and we should set it to bottom
-    if (sym->type_version) {
-        sym_set_bottom(ctx, sym);
-        return false;
+    JitSymType tag = sym->tag;
+    switch(tag) {
+        case JIT_SYM_NULL_TAG:
+            sym_set_bottom(ctx, sym);
+            return false;
+        case JIT_SYM_KNOWN_CLASS_TAG:
+            if (sym->cls.type->tp_version_tag != version) {
+                sym_set_bottom(ctx, sym);
+                return false;
+            }
+            else {
+                sym->cls.version = version;
+                return true;
+            }
+        case JIT_SYM_KNOWN_VALUE_TAG:
+            Py_CLEAR(sym->value.value);
+            sym_set_bottom(ctx, sym);
+            return false;
+        case JIT_SYM_TUPLE_TAG:
+            sym_set_bottom(ctx, sym);
+            return false;
```

Comment:
These could potentially have the same version tag, right? Looking up a method or whatever on a constant value is a normal thing to do.